### ISSUE TITLE
Get a list of custom modules 

### ISSUE THREAD

**Issue by TrudAX** [d365collaborative/d365fo.tools#722]
I want to write a script that compiles everything custom and stops on error. How to do this?
Previously I used the following script
```
Enable-D365Exception
Get-D365Model -CustomizableOnly -ExcludeMicrosoftModels -ExcludeBinaryModels | Invoke-D365ModuleCompile | Get-D365CompilerResult -OutputAsObjects
```
But it has several issues: 
1) It returns models, not modules
2) It doesn't list models in the correct order
I also found Get-D365Module, but it doesn't have  "-CustomizableOnly] [-ExcludeMicrosoftModels]-ExcludeBinaryModels" parameters

**Comment by FH-Inway** [comment]
[Get-D365Module](https://github.com/d365collaborative/d365fo.tools/blob/d739b376ad43cc4f49e2eb572298e611a05e1cf5/d365fo.tools/functions/get-d365module.ps1) does have the `-ExcludeBinaryModules` parameter. It also has the `-InDependencyOrder` parameter, which I'm assuming is the "correct order" you are after?

The idea is to first get the non binary modules with `Get-D365Module` in the right order. Then check with `Get-D365Module` using the `-Module`, `-ExcludeMicrosoftModels` and `-CustomizableOnly` parameters if the module is one you want to compile.

Not sure if this could be done in one command/pipe, but otherwise, it should work.

**Comment by FH-Inway** [comment]
Got a chance to play around with this. It works quite nicely and can be done in one line like this:

``` powershell
Get-D365Module -ExcludeBinaryModules -InDependencyOrder | Get-D365Model -ExcludeMicrosoftModels -CustomizableOnly | Invoke-D365ModuleCompile | Get-D365CompilerResult -OutputAsObjects
```

@TrudAX Let us know if this works for you.

**Comment by TrudAX** [comment]
Hi. Thanks for the reply.
I just ran 2 commands and it produced the wrong result 
Get-D365Module -ExcludeBinaryModules -InDependencyOrder | Get-D365Model -ExcludeMicrosoftModels -CustomizableOnly
e.g. I don't want to compile all sub..models of DEVTools

![image](https://user-images.githubusercontent.com/26786918/234477496-d65f0f00-f916-456a-bf77-ab504889281e.png)

**Comment by FH-Inway** [comment]
You can use the standard PowerShell cmdlet [Select-Object](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/select-object?view=powershell-5.1) with its `-Unique` parameter to get a list of the distinct module names. This list can then be piped to the `Invoke-D365ModuleCompile` cmdlet, which will compile each module only once, even if the module has more than one model.

``` powershell
Get-D365Module -ExcludeBinaryModules -InDependencyOrder | Get-D365Model -ExcludeMicrosoftModels -CustomizableOnly | Select-Object -Property Module -Unique
```

Side note: I tested this with the modules in the https://github.com/TrudAX/XppTools repo. The compilation of DEVTutorial fails with compile errors on a 10.0.34 environment (maybe on earlier versions as well). Also note that the modules need to be compiled at least once, otherwise `Get-D365Module` cannot discover them, as it uses the metadata runtime provider, which works on the compiled binaries.

**Comment by TrudAX** [comment]
Thank you for you help. I don't have .34 available, what are the compile errors?
The provided code works. Just one question, I want to write a script, that can be used on DEV VM after code refresh from DevOps(quite a common task). Is this the correct syntax for this: (the code itself seems working) or there is a better way?
`
$modules = Get-D365Module -ExcludeBinaryModules -InDependencyOrder | Get-D365Model -ExcludeMicrosoftModels -CustomizableOnly | Select-Object -Property Module -Unique
    $modules | Invoke-D365ModuleCompile | Get-D365CompilerResult -OutputAsObjects
    Invoke-D365DBSync -ShowOriginalProgress
    foreach ($model in $modules)
    {
        Invoke-D365ProcessModule -Module $model.Module  -ExecuteDeployReports 
    }
`

**Comment by FH-Inway** [comment]
DEVTutorial compile errors: They are related to the [Financial tags](https://learn.microsoft.com/en-us/dynamics365/release-plan/2023wave1/finance-operations/dynamics365-finance/define-financial-tags-support-analytical-reporting-business-rule-automation) feature that I believe was introduced with 10.0.32:
> On view 'LedgerJournalEntity', the field 'FinTagDisplayValue' refers to a nonexistent table or view named 'FinTag'.
> On view 'LedgerJournalEntity', the field 'OffsetFinTagDisplayValue' refers to a nonexistent table or view named 'FinTag'.

Regarding your script, looks good to me. Since you are already using [Invoke-D365ProcessModule](https://github.com/d365collaborative/d365fo.tools/wiki/Invoke-D365ProcessModule), you could simplify a bit and use it for compilation and synchronization as well. This has also the benefit that the compilation will include labels and reports compilation, which is not part of `Invoke-D365ModuleCompile`. If you don't want to use `Invoke-D365ProcessModule` for compilation, you should use `Invoke-D365ModuleFullCompile` instead.

``` powershell
$modules = Get-D365Module -ExcludeBinaryModules -InDependencyOrder | Get-D365Model -ExcludeMicrosoftModels -CustomizableOnly | Select-Object -Property Module -Unique
foreach ($module in $modules)
{
    Invoke-D365ProcessModule -Module $module.Module -ExecuteCompile -ExecuteSync -ExecuteDeployReports -ShowOriginalProgress
}
```

**Comment by TrudAX** [comment]
Hi. The task I tried to resolve is the following: I executed a get latest on my VM and want to prepare the app.
Of course, I want all labels and reports compiled and any errors reported. 

I tried your code, and it doesn't detect any compile errors(I executed Enable-D365Exception before it). For example, if there is an error in the module - I expect the execution stops, and the module will be reported, but now it just continues compiling. No indication to the user that something is wrong

Any advice on how we can stop the execution if any error is detected?  (this is the default behaviour of Invoke-D365ModuleCompile, but as you wrote, it doesn't compiles reports)

**Comment by FH-Inway** [comment]
Interesting behavior you found there. I tried various versions and was first stumped, because none of them stopped the compilation. Turns out that the `-ShowOriginalProgress` switch kinda hides errors. After removing it, I got the expected errors and together with `Enable-D365Exception` the script stopped execution at the first module with compile errors. This was the behavior for all 3 cmdlets that can be used to do a compilation (which makes sense: if it works for `Invoke-D365ModuleCompile`, it also must work for `Invoke-D365ModuleFullCompile` and `Invoke-D365processModule`, because internally they just call `Invoke-D365ModuleCompile`).

Turns also out my understanding of `Enable-D365Exception` was incomplete. As I documented in [Exception handling](https://github.com/d365collaborative/d365fo.tools/wiki/Exception-handling), my understanding was that `Enable-D365Exception` only works with cmdlets that implement the `-EnableException` switch. This is clearly not the case, as none of the 3 cmdlets for compilation implement that switch, but their behavior still changes.

Try the following version of the script:

``` powershell
Enable-D365Exception
$modules = Get-D365Module -ExcludeBinaryModules -InDependencyOrder | Get-D365Model -ExcludeMicrosoftModels -CustomizableOnly | Select-Object -Property Module -Unique
foreach ($module in $modules)
{
    Invoke-D365ProcessModule -Module $module.Module -ExecuteCompile -ExecuteSync -ExecuteDeployReports
}
```

**Comment by TrudAX** [comment]
ok, great. this seems working fine. Thanks for your help. The script I use is the following (I still prefer a full sync, as the DB may be a different version, it is safer)

```
    $modules = Get-D365Module -ExcludeBinaryModules -InDependencyOrder | Get-D365Model -ExcludeMicrosoftModels -CustomizableOnly | Select-Object -Property Module -Unique
    foreach ($model in $modules)
    {
        Invoke-D365ProcessModule -Module $model.Module  -ExecuteCompile 
    }
    Invoke-D365DBSync -ShowOriginalProgress
    foreach ($model in $modules)
    {
        Invoke-D365ProcessModule -Module $model.Module  -ExecuteDeployReports 
    }
```

**Comment by FH-Inway** [comment]
Looks good. Thanks to you as well, I got to learn more about the d365fo.tools cmdlets.

I think we can close this issue now, but feel free to ask additional questions or open a new issue or discussion.
